### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive with some union types

### ISSUE THREAD

**Issue by cainlevy** [typescript-eslint/typescript-eslint#10783]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&fileType=.tsx&code=C4TwDgpgBAqgdgSwPZygXigZ2AJwXAcygB8o4BXAWwCMIcBuAWACgWAzcuAY2GVTaRIAFAEoAXLEQooAbxZQoOCMHI5UAcnVNmAXxZcU2KAEN0UAcJHbTxzFlz4C9KAHoXUJAGsW%2Bw8CjUEvB8Zpra1CZ22HiEzm5kSP5eLEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

### Repro Code

```TypeScript
type Union = string | number;

function foo(): Union {
  return '';
}
const a = foo();
a as string; // ok

const b: Union = '';
b as string; // not ok
```

### ESLint Config

```javascript
module.exports = {
  "parser": "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect that `b as string` is considered to change the type of `b` by narrowing, as `a as string` does before it.

### Actual Result

`b as string` is tagged as an unnecessary type assertion

**Comment by kirkwaiblinger** [comment]
Hi @cainlevy! Turns out this is working as intended. TS narrows the type of `b` when you initialize a `const` variable regardless of the type annotation!

See [playground](https://typescript-eslint.io/play/#ts=5.7.3&fileType=.tsx&code=C4TwDgpgBAqgdgSwPZygXigZ2AJwXAcygB8o4BXAWwCMIcBuAWACgWBjFbKagLlkRTooAcmFNWzAGbk4bYMlTAAhgGsImAMq58BABQAPPtjyEAlFADeLFgF9rzAPQOoAFQAWCTFE9QlAGz8kAHcIABMAQhZlNU1tQl1qU3ogA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false)
```ts
type Union = string | number;

const b: Union = '';

function takesString(x: string) {
}

// This is allowed!
takesString(b);
```

If you dig through the no-unnecessary-type-assertion issues, you'll find several similar conversations.

**Comment by cainlevy** [comment]
@kirkwaiblinger Ok, thank you for taking the time to respond!
